Add ppc64le support for travis-ci

rff opened the pull request:
I know the community has being expecting the ppc64le travis integration, so I wanted to send this early patch to review and gather feedback.

There was a surprisingly small amount of changes to make this work, considering that was necessary to bump the distro version. On the commit messages there is more explanation about the details.

The purpose of this PR is, as the prefix implies, gather comments about this support and see if I am in the right direction. You will notice on the ci result that, if everything goes according to plan, all amd64 builds should succeed and some ppc64le too, but It is not complete, some some jobs for ppc64le is failing, but I want to get this out now for some early feedback.

I think theoretically, if the build-test for amd64 is really working as expected, that we can merge some of the changes, or maybe all with exception of the line in .travis.yml to add another arch, all the rest should work, even the custom ppc64le variables on travis.yml.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/4467)
<!-- Reviewable:end -->

rff commented:
@marcinslusarz, @pbalcer, in light of PR #4565 , what you would like to do about ppc64le CI support? GHA seems, AFAIK, to only support x86_64. I can re-enable all the travis-ci jobs for ppc64le only, If PMDK plan to keep travis-ci enabled in the long term.

marcinslusarz commented:
We are migrating to GHA because it's much faster and basically free, while Travis is quite expensive (only 5 jobs are free, and we are paying for 15). This doesn't mean we'll drop Travis completely - older branches (where most bug fixes go into) will still use Travis and for some repos migration is not worthwhile. 

With respect to the port, I think we can use Travis for PPC64. Most of the code in PMDK is shared, so I think we don't need as many combinations for PPC64 - leaving one coverage and one normal build should be enough.

What do you think?

BTW, I'm still waiting for this PR to build...

rff commented:
I updated the PR  with the following changes:

- Rebased on master, with that
  - removed the chown, it was implemented already
  - changed how to setup docker tty config
- Updated blacklist with data collected from travis builds
- Split the job list in two parts, the normal one for all archs and one for ppc64le only.
- Change valgrind version on ppc64le docker images 

This is now a working version, and I removed the [RFC] tag. I am specially looking forward for feedback about the valgrind approach I used and if installing from distro is ok or maybe I should get the upstream source). Also, if a ppc64le docker image version should be submitted to dockerhub.

@marcinslusarz 
> With respect to the port, I think we can use Travis for PPC64. Most of the code in PMDK is shared, so I think we don't need as many combinations for PPC64 - leaving one coverage and one normal build should be enough.
> What do you think?

LGTM. If some can be skipped, I Think it is OK, but I see that at least a half the jobs touch different paths during the build that expose architecture differences, like the packaging steps.
Anyway, we can always add more later if needed.

> BTW, I'm still waiting for this PR to build...

Sorry it took  a while. Now it seems fine.

ldorau commented:
@rff Is the `common: WIP: update Travis to Ubuntu 18.04.3 LTS Bionic` commit still WIP?

rff commented:
> @rff Is the `common: WIP: update Travis to Ubuntu 18.04.3 LTS Bionic` commit still WIP?

No, sorry, I just forgot It on the title. I will way to fix this together with other changes.

ldorau commented:
@rff please rebase, there is a merge conflict now

rff commented:
> @rff please rebase, there is a merge conflict now
Done, @ldorau, also removed "WIP" from that commit message.

Also removed obj_defrag_advanced from the blacklist, now that it meets the timeout requirement with -Og

marcinslusarz commented:
1) Installing Valgrind from distro packages on PPC is fine, but please pass distro name from Dockerfile to install-valgrind.sh and decide whether to use dnf or apt based on that.
2) We need per-architecture Docker images on DockerHub. With your current approach, I think Docker images will be rebuilt every time. We may need to expand the tag to include architecture in addition to OS and OSVER.
3) Please limit the number of jobs for PPC. I think these configurations should be enough:
  - FAULT_INJECTION=1 TEST_BUILD=debug
  - OS=fedora OS_VER=31 PMDK_CC=clang PMDK_CXX=clang++ TEST_BUILD=nondebug PUSH_IMAGE=1
4) Without "-t" the test framework doesn't use the colored output (PASS/FAIL messages). I'm not sure what -i is for.
5) I don't like the $arch__ env var support, because it might affect other variables. Please set BLACKLIST_FILE inside of build-travis.sh
6) I think the logic for deciding the value of TTY var is too complex. I think something like this should be used:
```
 if (travis) {
    if (ppc)
       ...
    else
       ...
 } else if (gha) {
   ...
 } else {
   ...
 }
```

marcinslusarz commented:
>     3. Please limit the number of jobs for PPC. I think these configurations should be enough:
>     * FAULT_INJECTION=1 TEST_BUILD=debug
>     * OS=fedora OS_VER=31 PMDK_CC=clang PMDK_CXX=clang++ TEST_BUILD=nondebug PUSH_IMAGE=1

Actually both jobs need PUSH_IMAGE=1

marcinslusarz commented:
Oh, btw, last commit message says it blacklists obj_defrag_advanced, but it's not true.

rff commented:
> 1. Installing Valgrind from distro packages on PPC is fine, but please pass distro name from Dockerfile to install-valgrind.sh and decide whether to use dnf or apt based on that.

Done.

> 2. We need per-architecture Docker images on DockerHub. With your current approach, I think Docker images will be rebuilt every time. We may need to expand the tag to include architecture in addition to OS and OSVER.

I am not that familiar with docker and dockerhub, but It is my understanding that dockerhub now supports metadata with the architecture and the workaround to specify the architecture on the image tag is not necessary anymore. For an example I run ubuntu on some initital travis testes for ppc64le and the command line was just:
`docker pull ubuntu` and docker would pull the image for the correct architecture.

What I think docker/dockerhub is missing is that if there is not a image for the host arch it just downloads a x86 one. That or this multi-arch support and checks are only used if the image uses manifest v2 ? In any case, that is why I had to put a manual architecture check. 
 
But back to the problem at hand, I do think it would be nice to push a ppc64le image to pmem/pmdk dockerhub. At least in theory we don't need to expand the tag for that, but I don't know the steps that are needed to make it happen.

>  3. Please limit the number of jobs for PPC. I think these configurations should be enough:
> * FAULT_INJECTION=1 TEST_BUILD=debug
> * OS=fedora OS_VER=31 PMDK_CC=clang PMDK_CXX=clang++ TEST_BUILD=nondebug PUSH_IMAGE=1
 
Done. But maybe should we keep the MAKE_PKG for fedora and ubuntu?
 
> 4. Without "-t" the test framework doesn't use the colored output (PASS/FAIL messages). I'm not sure what -i is for.

I just know what the manual says:
>  `--interactive` , `-i`      Keep STDIN open even if not attached

Maybe it is more useful for the `build-local.sh` and was only copied over. But IIRC `-t`/`--tty` needs `-i` to work, but I am not finding this on the docker manual now.  
I don't know if we can have a summary of what tests run or were skipped. But in my tests they seemed to run fine without any of this options on amd64.
 
> 5. I don't like the $arch__ env var support, because it might affect other variables. Please set BLACKLIST_FILE inside of build-travis.sh

Done.

> 6. I think the logic for deciding the value of TTY var is too complex. I think something like this should be used:

Done.

rff commented:
@marcinslusarz @ldorau 

Just pushed a new version, rebased on top of master and hopefully I addressed all comments without missing any.

rff commented:
https://travis-ci.org/pmem/pmdk/jobs/658223964

> time="2020-03-04T14:02:48Z" level=debug msg="[hijack] End of stdout"

During my tests I have seen this type of intermittent failure. If someone can restart the job I think It will succeed.

I don't know what causes it, but it seems to me that it is some interferences on the travis-ci itself, like when it cannot donwload the images or packages during the docker image setup. 

For reference, the same commit passed on other run, like here: https://travis-ci.com/rff/pmdk/builds/151747879

marcinslusarz commented:
> https://travis-ci.org/pmem/pmdk/jobs/658223964
> 
> > time="2020-03-04T14:02:48Z" level=debug msg="[hijack] End of stdout"

This is the actual reason:
make[4]: *** [../../Makefile.inc:78: clean-progs] Segmentation fault

This looks like a bug in the base system.

marcinslusarz commented:
> > 1. Please limit the number of jobs for PPC. I think these configurations should be enough:
> > 
> > 
> > * FAULT_INJECTION=1 TEST_BUILD=debug
> > * OS=fedora OS_VER=31 PMDK_CC=clang PMDK_CXX=clang++ TEST_BUILD=nondebug PUSH_IMAGE=1
> 
> Done. But maybe should we keep the MAKE_PKG for fedora and ubuntu?

Right

These 2 should be quite quick:
 * MAKE_PKG=1 REMOTE_TESTS=0 VALGRIND=0
 * MAKE_PKG=1 REMOTE_TESTS=0 VALGRIND=0 OS=fedora OS_VER=31

marcinslusarz commented:
@ldorau tested this and it seems DockerHub can't store 2 images with the same tag, but different architecture - second push has overwritten the first one.
@rff: do you think we should merge #4634 or is there another solution?

rff commented:
> @ldorau tested this and it seems DockerHub can't store 2 images with the same tag, but different architecture - second push has overwritten the first one.
> @rff: do you think we should merge #4634 or is there another solution?

@marcinslusarz I think there is another solution. Looking at the tags for the ubuntu image
https://hub.docker.com/_/ubuntu?tab=tags
and at this doc here
https://docs.docker.com/registry/spec/manifest-v2-2/#example-manifest-list

It suggests that you can define a manifest with multiple architectures with the same tag.

rff commented:
> Yes, it is possible indeed, but:
> 
>     1. Managing Docker image manifests and manifest lists is **experimental** on the Docker client and it **should not be used** in production environments:
>        https://docs.docker.com/engine/reference/commandline/manifest/

Thanks for the explanation @ldorau. I wasn't aware of this.

>     2. Adding managing Docker image manifests and manifest lists to PMDK's util/docker scripts would add a lot of unnecessary complications.
>
> IMHO the solution proposed in #4634 is much simpler and much more understandable and clear.

I agree.

marcinslusarz commented:
Ok, I've merged #4634.

marcinslusarz commented:
@ldorau GHA builds failed again with the same error message (```##[error]fatal: reference is not a tree: 5586724c69bb5ef281c2b8ae6a23033fcd60c6d2```) in "Clone the git repo" phase. Can you take a look?

ldorau commented:
> @ldorau GHA builds failed again with the same error message (`##[error]fatal: reference is not a tree: 5586724c69bb5ef281c2b8ae6a23033fcd60c6d2`) in "Clone the git repo" phase. Can you take a look?

@rff The above error is most likely due to the fact that GHA is confused by many branches being pushed with the same commit. You have three branches pointing to the same commit now: 
1) travis-ppc64le
2) travis-ppc64le-v5
3) wip-travis

This PR was opened with the 'travis-ppc64le' branch, but (for unknown reason - it can be a bug of GHA) there is no such branch among the GHA builds, but GHA CI builds your PR from the 'wip-travis' branch now:
https://github.com/pmem/pmdk/actions?query=branch%3Awip-travis

@rff Could you try to checkout the 'travis-ppc64le' branch, amend the last commit with resetting the date and push it again? (and please do NOT push the other branches with the same commit until all CI jobs clone your repo):
```
$ git checkout travis-ppc64le
$ git commit -a --amend --no-edit --reset-author
$ git push origin -f HEAD
```

ldorau commented:
> That is really weird.
> Since I had a little change do make anyway I did that and rebased on top of master. Also did not push to any other branch beside `travis-ppc64le` and just to make sure I did the `git commit -a --amend --no-edit --reset-author` on the top commit for the branch before push it.
> 

The root cause is probably different, see: https://github.com/pmem/pmdk/pull/4647

ldorau commented:
The Travis x86_64 Coverage build sometimes fails. I have restarted it.

marcinslusarz commented:
The failure on x86_64 is unrelated to this PR. It's just that when Docker image is rebuilt we go over the 50 minute limit. This is something we have to resolve, but not necessarily in this PR.
